pkg/upgrade/upgrades/alter_sql_instances_binary_version_test.go

```diff
@@ -0,0 +1,146 @@
+// licenses/APL.txt.
+
+package upgrades_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/server"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catenumpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/catconstants"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/upgrade/upgrades"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+)
+
+func TestAlterSystemSQLInstancesTableAddBinaryVersion(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			Knobs: base.TestingKnobs{
+				Server: &server.TestingKnobs{
+					DisableAutomaticVersionUpgrade: make(chan struct{}),
+					BinaryVersionOverride: clusterversion.ByKey(
+						clusterversion.V23_1AlterSystemSQLInstancesAddBinaryVersion - 1),
+				},
+			},
+		},
+	}
+
+	var (
+		ctx = context.Background()
+
+		tc    = testcluster.StartTestCluster(t, 1, clusterArgs)
+		s     = tc.Server(0)
+		sqlDB = tc.ServerConn(0)
+	)
+	defer tc.Stopper().Stop(ctx)
+
+	var (
+		validationSchemas = []upgrades.Schema{
+			{Name: "binary_version", ValidationFn: upgrades.HasColumn},
+		}
+	)
+
+	// Inject the old copy of the descriptor.
+	sqlInstancesTable := systemschema.SQLInstancesTable()
+	upgrades.InjectLegacyTable(ctx, t, s, sqlInstancesTable, getDeprecatedSqlInstancesDescriptorWithoutBinaryVersion)
+	// Validate that the table sql_instances has the old schema.
+	upgrades.ValidateSchemaExists(
+		ctx,
+		t,
+		s,
+		sqlDB,
+		keys.SQLInstancesTableID,
+		systemschema.SQLInstancesTable(),
+		[]string{},
+		validationSchemas,
+		false, /* expectExists */
+	)
+	// Run the upgrade.
+	upgrades.Upgrade(
+		t,
+		sqlDB,
+		clusterversion.V23_1AlterSystemSQLInstancesAddBinaryVersion,
+		nil,   /* done */
+		false, /* expectError */
+	)
+	// Validate that the table has new schema.
+	upgrades.ValidateSchemaExists(
+		ctx,
+		t,
+		s,
+		sqlDB,
+		keys.SQLInstancesTableID,
+		sqlInstancesTable,
+		[]string{},
+		validationSchemas,
+		true, /* expectExists */
+	)
+}
+
+// getDeprecatedSqlInstancesDescriptorWithoutBinaryVersion returns the system.sql_instances
+// table descriptor that was being used before adding a new column in the
+// current version.
+func getDeprecatedSqlInstancesDescriptorWithoutBinaryVersion() *descpb.TableDescriptor {
```

Comment:
It's possible to bootstrap the DB with the V22_2 system database. It makes writing upgrade tests a lot easier and the tests are more representative of how the upgrades will run in production. See https://github.com/cockroachdb/cockroach/pull/98613 for an example.

Reply:
Done